snuba/query/processors/mapping_optimizer.py

```diff
@@ -159,20 +200,111 @@ def __replace_with_hash(self, condition: Expression) -> Expression:
             ),
         )
 
+    def _get_condition_without_redundant_checks(
+        self, condition: Expression, query: Query
+    ) -> Expression:
+        """Optimizes the case where the query condition contains the following:
+
+        valueOf('my_tag') != '' AND valueOf('my_tag') == "something"
+                          ^                            ^
+                          |                            |
+                      existence check               value check
+
+        the existence check in this clause is redundant and prevents the hashmap
+        optimization from being applied.
+
+        This function will remove all tag existence checks
+        from the condition IFF they are ANDed with a value check for the *same tag name*
+
+        Side effects:
+            This function works by flattening first level AND conditions to find clauses where
+            existence checks and value checks are ANDed together. When the AND conditions are recombined,
+            they are not guaranteed to be in the same structure (but are guaranteed to be functionally equivalent)
+
+            Example:
+                ┌───┐         ┌───┐
+                │AND│         │AND│
+                ├──┬┘         └┬──┤
+                │  │           │  │
+             ┌──┴┐ c           a ┌┴──┐
+             │AND│    becomes    │AND│
+             └┬─┬┘               ├──┬┘
+              │ │                │  │
+              a b                b  c
+        """
+        # how often we should skip this optimization
+        skip_rate = state.get_config("tags_redundant_optimizer_skip_rate", 0)
```

Comment:
This function is recursive, so this might get called multiple times. I would suggest fetching this value in the `process_query` function and passing it in to this function.

Reply:
good point. Actually even better we this whole function can be skipped outside of itself